### DISCUSSION TITLE
rust-course/basic/method

### DISCUSSION THREAD

**Discussion by sunface** [sunface/rust-course#635] [章节评论区]
https://course.rs/basic/method.html

**Comment by zhangairku** [comment]
这里是不是少讲了函数指针，fn类型

**Reply by zrll12** [reply]
Fn类型在进阶的闭包里讲

**Comment by Brankond** [comment]
pub struct Rectangle {
    width: u32,
    height: u32,
}

impl Rectangle {
    pub fn new(width: u32, height: u32) -> Self {
        Rectangle { width, height }
    }
    pub fn width(&self) -> u32 {
        return self.width;
    }
}

fn main() {
    let rect1 = Rectangle::new(30, 50);

    println!("{}", rect1.width());
}

这个位置为什么说 rect1.width 是私有的呢？
使用 pub 定义了 Rectangle 结构体后 width 字段不是应该变成 public 了吗？

**Reply by sendking** [reply]
这里试了一下是可以访问 rect1.width ，并没有觉得是私有的

**Reply by PaiGack** [reply]
在本文件内部应该是可以访问了 （相当于 保护），但是在别的 mod 里面就没发直接访问的

**Reply by zero-edge-code** [reply]
越到后面评论区人越少啊

**Reply by ramonli** [reply]
哈哈，越到后面 人越少

**Reply by kikohz** [reply]
从入门到放弃吗 😓

**Reply by noo-have** [reply]
默认所有字段都是私有的，只对当前文件公开；

**Reply by Neisseri** [reply]
那如果不加public的话在当前文件里能访问吗

**Reply by arronKler** [reply]
在 mod 中对外导出的时候，外部会无法访问这个 width，但对于和 mod 自身同层级的是可见的。如果是对 enum 设置 pub，则其中的每个值就默认都是 public，无需单独每个都设置一下 pub

**Reply by bluecode2017** [reply]
我理解是一旦有了公开的get 方法比如Width（）， 那字段就默认变成私有的。

**Reply by eatwellw** [reply]
加pub在模块内是公开的，在模块外是私有的。
`
mod my_module {
    // 在 my_module 模块中定义一个私有结构体
    struct PrivateStruct {
          field: i32,
    }
    impl PrivateStruct {
        // 定义一个公共方法，用于修改私有字段的值
        pub fn set_field(&mut self, value: i32) {
            self.field = value;
        }
    }
}
fn main() {
    // 创建一个 my_module 模块的作用域
    {
        use my_module::PrivateStruct;
        let mut struct_instance = PrivateStruct { field: 42 };
        // 通过公共方法修改私有字段的值
        struct_instance.set_field(24);
        // 输出私有字段的值
        println!("私有字段的值为: {}", struct_instance.field);
    }
}
`
这个代码汇报错；在 struct PrivateStruct前加上pub，在字段加上pub才能让这个代码运行。

**Reply by eatwellw** [reply]
mod my_module {
    // 在 my_module 模块中定义一个私有结构体
    struct PrivateStruct {
          field: i32,
    }
    impl PrivateStruct {
        // 定义一个公共方法，用于修改私有字段的值
        pub fn set_field(&mut self, value: i32) {
            self.field = value;
        }
    }
}
fn main() {
    // 创建一个 my_module 模块的作用域
    {
        use my_module::PrivateStruct;
        let mut struct_instance = PrivateStruct { field: 42 };
        // 通过公共方法修改私有字段的值
        struct_instance.set_field(24);
        // 输出私有字段的值
        println!("私有字段的值为: {}", struct_instance.field);
    }
}

**Comment by wentaotoyou** [comment]
只要是定义在impl中的函数就叫关联函数吧？而不管有没有self参数

**Reply by Yellster** [reply]
我的理解是，有self参数，就不是函数，而是方法了

**Reply by wentaotoyou** [reply]
先看本教程的，看其它材料时发现这个概念冲突了，才查了英文版的： https://doc.rust-lang.org/book/ch05-03-method-syntax.html#associated-functions
不知道是不是我理解有误

**Reply by Aresona** [reply]
We can define associated functions that don’t have self as their first parameter (and thus are not methods) 感觉这里的意思应该是没有 self 作为第一个参数时是关联函数(associated function), 而不是方法(method)

**Reply by GmwEnterprise** [reply]
「Unlike functions, methods are defined within the context of a struct (or an enum or a trait object, which we cover in Chapters 6 and 17, respectively), and their first parameter is always self」这是英文版本里提到的，文档可以通过命令 `rustup doc ` 来查找，具体本地路径：`.rustup/toolchains/stable-x86_64-pc-windows-msvc/share/doc/rust/html/book/index.html`

**Reply by ghost** [reply]
很明确，第一个参数是`self`或者`&self`的就是方法，否则就是关联函数，这种差异直接体现在调用方法上：
关联函数：可以独立调用，无需类型的实例，因为毕竟本身就是个函数。但由于关联于某个类型，所以得使用类似`String::from()`的方式；
方法：无法独立调用，必须先有类型的实例，再有方法。比如f.read()等。

**Reply by doge-ak** [reply]
关联函数有点像类的静态方法，方法类似于成员方法

**Reply by bluecode2017** [reply]
Rust语言新的概念好多啊，感觉绕的很

**Reply by Clazex** [reply]
可以看出有些人把答案怼面前了还是只会瞎编，甚至就在同个页面的同一个和下一个自然段都不愿看：

> We *can* define associated functions that don’t have `self` as their first parameter ...

言下之意：关联函数*也可以*有 `self` 参数。

> Associated functions *that aren’t methods* are often used for ...

言下之意：*有的*关联函数是方法。



实在不能理解什么叫“……的就是方法，否则就是关联函数”，《很明确》。

方法 *methods* 是关联函数 *associated functions* 的子集。只要在 `impl` 块中的都叫 *associated functions*，*其中*，有 `self` 参数的是 *methods*。

与 *associated functions* 对应的概念是 *free functions* 自由函数，即不在 `impl` 块中的函数，而不是方法。

**Reply by ysy** [reply]
https://doc.rust-lang.org/reference/items/associated-items.html

*Associated functions whose first parameter is named self are called methods and may be invoked using the method call operator, for example, x.foo(), as well as the usual function call notation.*

很显然方法(Method)是关联函数(Associated functions)的子集。教程中的表述： `这种定义在 impl 中且没有 self 的函数被称之为关联函数`     是错误的，希望可以改下。

**Reply by duanyuluo** [reply]
Function (Unassociated with type) -> fn main() {}
Asscoiated Functons
-- method (of a object instance) -> fn area(&self) -> usize {}
-- other asscoiated function(of a type) -> impl Rect { fn from(...) {} }

**Reply by duanyuluo** [reply]
Asscoiated with some type(eg. struct, enum, trait...)

**Reply by Bruce-Wall** [reply]
确实，关联函数包含方法和第一参数不是self的关联函数，

作者可能觉得第一参数不是self的关联函数也是关联函数，
并且方法第一参数是self，第一参数不是self的不是方法也没毛病。

单看这两点倒是没毛病，但是文中这种表述有二义性，
这种第一参数不是self的是关联函数没毛病，所以单这句话其实没毛病，但是后面不说方法也属于关联函数的话，这里就可以有关联函数仅指第一参数不是self的impl中函数的语义，这个歧义存在且把关联函数的定义范围缩小了。

楼上有的老哥发言倒也不至于这么尖酸刻薄，别人都是有错指错就完事了。

不过作者到现在依然还没改确实不应该，希望作者能尽快把这里的定义修正。
虽然单看话本身从语文的角度没什么毛病，但不管是作者理解偏差还是表达问题，这里都会对读者产生误导，产生概念性错误，需要更正。

**Comment by lsm1998** [comment]
关联函数的描述，‘Rust 中有一个约定俗称的规则’，应该是约定俗成

**Reply by BiteFoo** [reply]
约定好了就遵守吧

**Comment by kingwingfly** [comment]
这个枚举类型好像Python里的类属性啊，或者`__init__`下面的东西哦。。。
还有那个`can_hold`不是很严谨，应该可以考虑旋转
```rust
impl Rectangle {	//为Rectangle结构体增加一种方法
    fn can_hold(self, other: Rectangle) -> bool {
        if (self.length >= other.length && self.width >= other.width)||(self.length >= other.width && self.width >= other.length) {
            return true
        }
        false
    }
}

**Reply by supernova4869** [reply]
Python的__init__下面是它的构造函数，类似这里的new

**Reply by ZLD1819858** [reply]
emm这里是讲用法，你这样考虑就有点轴了

**Reply by bluecode2017** [reply]
方法怎么定义和命名还是设计和写code的人说了算嘛

**Comment by 791045873** [comment]
let rect3 = Rectangle { width: 60, height: 45 };
 let sq = Rectangle::new(3, 3)；

这两种写法是等价的吗？
如果不等价有啥区别

**Reply by noo-have** [reply]
这里的new并非关键词，只是Rectangle结构体身上实现了的一个方法；如果未实现new方法自然不可调用。第一种写法是对所有结构体通用的。

**Reply by BiteFoo** [reply]
我的理解就是如果没有实现new方法，那么就用通用的结构体创建实例，因为rust没有默认给每个结构体设置默认的new方法，因此不能直接调用T::new(xxx)方法，

**Reply by bluecode2017** [reply]
“let rect3 = Rectangle { width: 60, height: 45 };”
这种用法是不是出了这个rs文件就不能这么写了？我记得说所有property是私有的嘛

**Reply by zhenzhenChange** [reply]
我觉得吧，这就相当于一个字面量形式和一个构造形式，结果是一样的，代入到 TS：
```ts
const r1: Rectangle = { width: 60, height: 45 }
const r2 = new Rectangle(60, 45)
```

**Reply by whtblktone** [reply]
这里展示的没有区别，但是有可能会在new函数中做些参数检查工作

**Reply by duanyuluo** [reply]
准确的说，你在Rectangle::new()中必须调用Rectangle{}才能构造出结构。

**Reply by duanyuluo** [reply]
new（）方法是你自己实现的，不是rust实现的。

**Reply by duanyuluo** [reply]
比如你可以不实现new，只写Rectangle::from()，就不会搞混了，你太熟悉new了。

**Reply by QIN2DIM** [reply]
如果你为 Rectangle 实现了 new 方法，你可以理解为 new 包含了实例化。一般而言，会在 new 中做除了返回结构体 Self 以外的事，比如初始化一些默认字段，做输入参数的校验和模式匹配等等。

**Reply by Gvhthon** [reply]
其结果而言是等价的。但第一个只是简单的创建了一个结构体的实例。而第二个是利用自定义的结构体的new方法创建实例，而在这个new方法中，可以而外的执行一些东西，比如初始化检测错误等等。

**Reply by nowethan** [reply]
```
pub struct Student {
    pub name: String,
    age: i32,
}

impl Student {
    pub fn new(name: String, age: i32) -> Student {
        Student { name, age }
    }
}

在其他mod 无法通过  let a:Student  = Student {} 来完成初始化，  可以通过  Student ::new()方法完成创建

```

**Comment by bluecode2017** [comment]
”Rust 中有一个约定俗成的规则，使用 new 来作为构造器的名称，出于设计上的考虑，Rust 特地没有用 new 作为关键字“
这个是什么意思呢，看不太懂。 是不是说new在其他语言是操作符，是关键字。而在这里，只是一个普通的构造函数/方法？

**Reply by bluecode2017** [reply]
好奇为什么不默认把每个struct都自动实现好一个new方法呢？还得自己手写。

**Reply by shenyongqing** [reply]
我的理解是rust的每一个变量都要求明确的值（不像go），如果结构体成员太多，struct {::::}写起来非常不方便，用一个new函数给出几个关键的成员值，其他就用默认值好了。那么制造一个对象就简洁多了。

**Reply by duanyuluo** [reply]
第一，在C中，alloc只负责分配内存，结构体的初始化还是要你自己写的。
第二，在Rust中，不推荐出现没有明确初始化的fields。
第三，所谓的默认实现的new都是伪的，因为你在真正使用前不初始化，是不可预测的。
第四，你可以参考Default trait，rust只是不隐藏的默认做而已。

**Comment by Vocal-map** [comment]
关联函数和方法的区别是啥？

**Reply by ZLD1819858** [reply]
关联函数：和某个结构体有关联的函数
方法：必须通过对象来调用的叫做“方法“
例如：
    let mut v1 = Vec::new(); //调用Vec上的关联函数，创建一个空的Vector
    v1.push(1);    // 调用v1对象上的push方法向v1对象中压入一个数据1

假设这样调用：Vec::push(1); 从语义上来讲就是不对的，简单翻译一下：向一个Vec中压入一个数据1，那么这个Vec代指的谁？堆上的哪个vec？所以方法需要通过对象来调用，即对象为主语：我（v1）压入一个数据1.

不知道我这样解释你能理解吗，你之前应该没有其它语言的基础，直接学的Rust吧？在Java中有”静态方法“和”方法“等同于Rust中的关联函数和方法。在C语言中也有类似的，直接调用的函数和成员函数（结构体中的存储函数指针）。

**Reply by Vocal-map** [reply]
明白明白，想起来了，C就了解到指针，C++只是拿来搞算法，只知道STL，其他时间都在Java。。突然看到个名词有点懵

**Reply by ka3hun9** [reply]
关联函数 : 相当于结构体的静态方法,
方法 : 相当于结构体的实例方法

**Reply by duanyuluo** [reply]
不必拘泥名词，只要记得
fn main()
impl XX { fn foo() {} } -> XX::foo()
impl XX {fn foo(&self) {} } -> obj.foo()
let f = |x| x+1; -> f(2);

function
method of type
method of object
lamda function

**Reply by QIN2DIM** [reply]
如果你用过 Python，可以这么理解。
**关联函数**：Python class 中的 *@staticmethod*，可以从 `ClassName.func()` 直接使用。这些“静态方法”不会通过 `self.x` 的方式访问类中的其他成员，他完全可以作为一个单独的函数放在其他地方，但只是说这些特殊的 func 和类的应用场景强相关，所以写到一起，便于维护，增加可读性。
**方法**：Python class 中的 *@classmethod*，必须先实例化 `instance = ClassName()` 才能调用 `instance.func()`

**Reply by JoshuaGhost** [reply]
大佬可能笔误了吧，方法不需要用@classmethod装饰器啦，而且@classmethod本来就支持直接用类名来调用，而且第一个参数cls是指代本类的。这也就是为啥工厂模式是用这玩意实现的。比如：

```python
Class Foo:
    @classmethod
    def foo_factory(cls, *args, **kwargs) -> "Foo":
        do_something()
        return cls(*args, **kwargs)
```

而方法就是那些有self不用两个装饰器的方法，比如

```python
Class Bar:
    def do_anything(self, buz) -> int:
        self.do_everything(buz)
        return 42
```

这个do_anything的函数是方法，就是只能通过instanciation来调用：

```python
bar = Bar()
assert bar.do_anything == 42
```

**Reply by QIN2DIM** [reply]
emm，感谢提醒 这里我没有表达清楚。@JoshuaGhost

**Reply by Gvhthon** [reply]
我的理解是如果你想在函数里调用实例的字段（属性），那就用方法。如果只是简单的实现一些动作，用关联函数就可以了。比如a.changeNumber(18);

**Reply by xbsheng** [reply]
看第一个参数是不是`&self`/`&mut self`/`self`就可以区分，是这几个值就代表是方法，不是就代表是关联函数。个人认为rust中的方法就是其他语言中的实例方法，关联函数就是静态方法，不确定这个理解确切不~

**Comment by duanyuluo** [comment]
ref rustdoc std::ops
Fn
FnMut
FnOnce
Only three types of functions
分别对应
不影响闭包环境
影响闭包环境
影响且只能一次影响闭包环境

https://blog.csdn.net/wowotuo/article/details/80785358#:~:text=Fn%EF%BC%9A%E5%AE%83%E5%8F%AA%E8%83%BD%E4%BB%8E,%E5%BC%95%E5%87%BAFn%EF%BC%8CFnMut%EF%BC%8CFnOnce%3F
这篇文章讲的很好

**Reply by duanyuluo** [reply]
发错地方了……应该在closure中。。。。

**Reply by duanyuluo** [reply]
不过我觉得，fn就是给一个闭包命名 ：D

**Reply by jieqiyue** [reply]
怎么哪里都有你

**Comment by tayuqitan** [comment]
这章两个概念比较绕，关联函数和方法的区别。这里做个笔记：
     1、最大的区别就是，关联函数绑定给类型。方法绑定给实例。
     2、关联函数和方法都可以使用obj[Obj]::fn方式调用。只是调用方法的时候需要传入&self参数。eg：Rectangle::width(&rect);
     3、关联函数不能通过obj.fn方式调用。只能通过Obj::fn方式调用。

**Comment by CListery** [comment]
关联函数，就叫构造函数又怎么了嘛，非要取些怪头怪脑的名字让人不便记忆

**Reply by CListery** [reply]
只是吐槽，没有任何其他意思

**Reply by tayuqitan** [reply]
其实叫静态函数，更贴切一些。但是也不全对，所有绑定给类型的都叫：关联函数。

**Reply by CListery** [reply]
不敢苟同，以下是个人见解：
1. 静态函数，一般指不需要构建对象的情况下就可以调用该对象的内部函数
2. 关联函数，在rust中指的是用于构建对象的特殊函数
3. 方法，在rust中指的是除关联函数外的函数

**Reply by Dmcz** [reply]
关联函数也并非用于构建对象

**Reply by Dmcz** [reply]
错了，应该是并非只用于构建对象...

**Reply by liuhq** [reply]
@CListery 错了，所有在 `impl` 块中实现的函数都叫关联函数，其中第一个参数是 Self 类型的关联函数可以称为方法，所以方法是关联函数的子集，而不是互斥的关系

**Comment by yanzongt** [comment]
关联函数比构造函数更灵活，还不用考虑c++中一系列的继承析构问题；
尤其枚举类型，不仅可以嵌套匹配，还可以定义方法，我们rust真的太强大啦。

**Comment by wslzs** [comment]
我理解 ‘&self’相当于java中的this

**Reply by xbsheng** [reply]
俺也一样

**Comment by madjxatw** [comment]
如果理解正确，对比C++，关联函数类似 `static` 成员函数，表示与某个类相关的函数；方法类似成员函数，作用于类的实例对象。

**Reply by liuhq** [reply]
理解错了，所有在 impl 块中实现的函数都叫关联函数，其中第一个参数是 Self 类型的关联函数可以称为方法，所以方法是关联函数的子集，而不是互斥的关系

**Comment by madjxatw** [comment]
教程中应该加入一些其他语言的类比，帮助加快理解速度。如果只是扣文字和一些简单案例，效率低，还容易造成困惑。

**Reply by xbsheng** [reply]
但是其他语言太多了，读者不一定熟悉类比的语言

**Comment by Rgoogle** [comment]
Rust 会自动为 object 添加 &、&mut 或 * 以便使 object 与方法签名匹配。 这句话是错误的，&mut这个不会加，会报错啊！在实现了copy trait的时候 好像又莫名其妙的能调用了，但是前提条件却没有

**Comment by Skyfacon** [comment]
学习打卡

**Comment by Datu1213** [comment]
大哥加油。

**Comment by TianChen-Yu** [comment]
所以自由函数、关联函数、方法这三者在底层实现上是怎样的？在运行和性能上有差异吗？

我的理解是，
- 所谓 关联函数 就是需要 `类型::函数名` 访问的函数（所以就是 *与类型关联* ）
- 所谓 方法 就是 在第一个参数使用 `self` 关键字 的关联函数（是关联函数的一种）

然而，Rust有个很灵活的特性：类型实例的 方法 实际上也可以通过 关联函数 的形式进行调用，
比如

```rust
// 定义所用的结构体
struct S(i32);
// 这是结构体的 实例方法
impl S {
    pub fn add_from(&mut self, other: i32) {
        self.0 += other;
    }
}
// 与之等价的 自由函数
pub fn add_from(s: &mut S, other: i32) {
    s.0 += other;
}

fn main() {
    // 初始化一个结构体
    let mut s = S(0);
    // 以 实例方法 的形式调用
    s.add_from(1);
    // 以 关联函数 的形式调用
    S::add_from(&mut s, 1);
    // 用 自由函数 调用
    add_from(&mut s, 1);
}
```
这里边的 关联函数 个人觉得已经很像 自由函数 了（除了加个类型作为命名空间区分），
实际编译的时候是否也会有 自由函数 的性能？
或者是否可以说，
实际上整个 关联函数 实例方法 的机制，
都是Rust 零成本抽象 的语法糖？

**Reply by qiyuewuyi2333** [reply]
自由函数和关联函数的性能没什么区别的。（来自微量的benchmarks）

**Comment by hamsterjiang23** [comment]
```rust
 // 可以为枚举实现方法
        enum Message{
            Quit {boolean : bool},
            Move {x : i32, y : i32},
            Write(String),
            ChangeColor(i32, i32, i32)
        }

        impl Message{
           fn call(&self) -> (){
            // 输出write里的string
            match self {
                Message::Write(string) => {
                    println!("Message string is {}", string)
                },
                Message::ChangeColor(x, y , z) => {
                    println!("Message change color is ({}, {}, {})", x, y, z)
                }
                _ => ()
            }
           }
        }

        let msg1 = Message::Write("Hello".to_string());
        let msg2 = Message::ChangeColor(1,2,3);
        msg1.call();
        msg2.call();
```

**Comment by feiandre** [comment]
Trait 其他教材翻译成特型, 这里翻译成特征. 感觉怪怪的

**Comment by Murasame-BH8HJY** [comment]
是不是我还没学到难点，至少目前觉得还挺简单的😉

**Comment by xbsheng** [comment]
这章类比其他语言的静态方法、实例方法学习比较容易理解~

**Comment by typed-sigterm** [comment]
> 在 C/C++ 语言中，有两个不同的运算符来调用方法：`.` 直接在对象上调用方法，而 `->` 在一个对象的指针上调用方法，这时需要先解引用指针。换句话说，如果 `object` 是一个指针，那么 `object->something()` 和 `(*object).something()` 是一样的。

其实 `->` 才是语法糖

**Comment by jkbody** [comment]
done

**Comment by samwu97** [comment]
补充个self的例子：
```rust
struct Rectangle {
    width: u32,
    height: u32,
}

impl Rectangle {
    // 值的所有权
    fn consume(self) {
        println!("Consuming Rectangle: {}x{}", self.width, self.height);
    }

    // 不可变引用
    fn area(&self) -> u32 {
        self.width * self.height
    }

    // 可变引用
    fn set_width(&mut self, width: u32) {
        self.width = width;
    }
}

fn main() {
    let mut rect = Rectangle { width: 30, height: 50 };

    // 使用不可变引用
    println!("Area: {}", rect.area());

    // 使用可变引用
    rect.set_width(40);
    println!("New Area: {}", rect.area());

    // 转移所有权
    rect.consume();
    // rect.area(); // 错误，因为所有权已被 consume() 方法转移
}

```

**Comment by safarishi** [comment]
接下来的内容非常重要 --- 这个文本建议加粗

**Comment by reatang** [comment]
所以【关联函数】就是【静态方法】呗？Rust每个概念都得改个名字，确实累。
